service/controller/resource/spark/create.go

```diff
@@ -0,0 +1,308 @@
+	if err := r.ctrlClient.Get(ctx, objectKey, cluster); err != nil {
+		return nil, err
+	}
+
+	r.logger = r.logger.With("cluster", cluster.Name)
+
+	return azureCluster, nil
+}
+
+// getClusterByName finds and return a Cluster object using the specified params.
+func (r *Resource) getClusterByName(ctx context.Context, namespace, name string) (*capiv1alpha3.Cluster, error) {
+	cluster := &capiv1alpha3.Cluster{}
+	objectKey := client.ObjectKey{Name: name, Namespace: namespace}
+	if err := r.ctrlClient.Get(ctx, objectKey, cluster); err != nil {
+		return nil, err
+	}
+
+	r.logger = r.logger.With("cluster", cluster.Name)
+
+	return cluster, nil
+}
+
+// getOwnerCluster returns the Cluster object owning the current resource.
+func (r *Resource) getOwnerCluster(ctx context.Context, obj metav1.ObjectMeta) (*capiv1alpha3.Cluster, error) {
+	for _, ref := range obj.OwnerReferences {
+		if ref.Kind == "Cluster" && ref.APIVersion == capiv1alpha3.GroupVersion.String() {
+			return r.getClusterByName(ctx, obj.Namespace, ref.Name)
+		}
+	}
+
+	return nil, nil
+}
+
+// getMachinePoolByName finds and return a MachinePool object using the specified params.
+func (r *Resource) getMachinePoolByName(ctx context.Context, namespace, name string) (*expcapiv1alpha3.MachinePool, error) {
+	machinePool := &expcapiv1alpha3.MachinePool{}
+	objectKey := client.ObjectKey{Name: name, Namespace: namespace}
+	if err := r.ctrlClient.Get(ctx, objectKey, machinePool); err != nil {
+		return nil, err
+	}
+
+	r.logger = r.logger.With("machinePool", machinePool.Name)
+
+	return machinePool, nil
+}
+
+// getOwnerMachinePool returns the MachinePool object owning the current resource.
+func (r *Resource) getOwnerMachinePool(ctx context.Context, obj metav1.ObjectMeta) (*expcapiv1alpha3.MachinePool, error) {
+	for _, ref := range obj.OwnerReferences {
+		if ref.Kind == "MachinePool" && ref.APIVersion == expcapiv1alpha3.GroupVersion.String() {
+			return r.getMachinePoolByName(ctx, obj.Namespace, ref.Name)
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *Resource) getRelease(ctx context.Context, obj metav1.ObjectMeta) (*releasev1alpha1.Release, error) {
+	release := &releasev1alpha1.Release{}
+	err := r.ctrlClient.Get(ctx, client.ObjectKey{Namespace: corev1.NamespaceAll, Name: key.ReleaseVersion(&obj)}, release)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return release, nil
+}
+
+func secretName(base string) string {
+	return base + "-machine-pool-ignition"
+}
+
+// toAzureConfig performs mapping from CAPI & CAPZ types to AzureConfig which
+// fulfills k8scloudconfig requirements for rendering node pool worker node
+// ignition templates.
+func toAzureConfig(cluster *capiv1alpha3.Cluster, azureCluster *capzv1alpha3.AzureCluster, machinePool *expcapiv1alpha3.MachinePool, azureMachinePool *expcapzv1alpha3.AzureMachinePool) providerv1alpha1.AzureConfig {
+	ac := providerv1alpha1.AzureConfig{}
+
+	// TODO: Implement <3
```

Comment:
This is pretty much just annoying struct construction by copying relevant fields from available CRs.

Reply:
I'm not fully sure if my grep finds everything, but there might not be _that_ many fields to set. Note that these are from [CommonClusterType](https://github.com/giantswarm/apiextensions/blob/master/pkg/apis/provider/v1alpha1/azure_types.go#L35):
```
k8scloudconfig $ grep -R '\.Cluster\.' | sed -s 's|.*\(\.Cluster[.a-zA-Z]*\).*|\1|g'|sort |uniq
.Cluster.Calico.CIDR
.Cluster.Calico.MTU
.Cluster.Docker.Daemon.CIDR
.Cluster.Etcd.Domain
.Cluster.Etcd.Prefix
.ClusterIPRange
.Cluster.Kubernetes.API.Domain
.Cluster.Kubernetes.API.SecurePort
.Cluster.Kubernetes.CloudProvider
.Cluster.Kubernetes.DNS.IP
.Cluster.Kubernetes.Domain
.Cluster.Kubernetes.Kubelet.Domain
.Cluster.Kubernetes.Kubelet.Labels
.Cluster.Kubernetes.NetworkSetup.KubeProxy.ConntrackMaxPerCore
.Cluster.Kubernetes.SSH.UserList
```